Build a basic Spring Boot Application

mdiwale opened the pull request:
<!--
## Submitting a Pull Request
We love PRs and appreciate any help you can offer!

Please make sure the following criteria are met before submitting your pull request.

1. <strong>PR meets the Contributing Guidelines (see above)</strong>
2. Ensure the test suite passes
3. Make sure your code lints
-->

## Related Issue
<!-- Include a link to the issue (e.g. #12) -->
[Setup meetup API repository with basic build and CI process #4](https://github.com/ProvidenceGeeks/website-api-events/issues/4)

## Summary of Changes
<!-- Briefly summarize the changes made, lists are also appreciated.  Referencing the related issue as a
"TO DO" checklist is also helpful for reviewers

1. [x] Thing I fixed
1. [x] Other thing I updated
1. [x] Docs I updated

--> 
1. Built a basic "Hello World" Spring Boot application
2. Updated README file with the project setup.

TO DO

1. Update the Wiki with local development environment setup as follows:

a. How to install Java 

Java development on any platform often benefits from the use of an Integrated Development Environment (IDE), which provides a more fluid workflow between writing, compiling, running, debugging, and packaging Java code than a simple text editor and the command line. Different IDEs offer unique features and are often suited for different kinds of Java development. These IDEs are industry leaders and offer substantial support for OS X:
* Eclipse IDE for Java Developers ([http://www.eclipse.org](http://www.eclipse.org)) is a free download.
* Netbeans IDE ([http://www.netbeans.org](http://www.netbeans.org)) is a free download.
* Jetbrains IntelliJ IDEA ([http://www.jetbrains.com/idea/](http://www.jetbrains.com/idea/)) requires a license for continued use after a trial period.
* Xcode ([http://developer.apple.com/tools/xcode/](http://developer.apple.com/tools/xcode/)) is a free download with a free account from the Apple Developer Connection.

b. How to install Mavin

Refer the below link to install Mavin:
[http://www.admfactory.com/how-to-install-maven-on-macos/
](http://www.admfactory.com/how-to-install-maven-on-macos/)

2. Update Wiki with the information about mvnw and mvnw.cmd files as follows:

These files are from Maven wrapper.
This allows you to run the Maven project without having Maven installed and present on the path. It downloads the correct Maven version if it's not found.
The mvnw file is for linux (bash) and the mvnw.cmd is for Windows environment.

thescientist13 commented on README.md:
```diff
@@ -4,6 +4,13 @@
 Backend RESTful API for managing event data to be consumed by clients, like *website-frontend*.  Built with Java and Spring Boot.
 
 ## Project Setup
+Please review the local development environment steps at * link to the section in the wiki * and after you have Java and Maven installed do the following to setup the project.
```
Hey @mdiwale , so far so good, but have a couple changes for this section:
1. "* link to the section in the wiki *" was meant to be a prompt to you to include a link to the Onboarding Documentation for that section.  Please update.
1.  Please use numbered listing for the steps, e.g.
> 1.  Step 1 
> 1.  Step 2
3.  Where does this project come from you reference in Step 1?
4. Any steps for installing dependencies?
5. How do I build or run tests?  An example command to validate everything is working might be good.

thescientist13 commented on test.iml:
```diff
@@ -0,0 +1,74 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Can we confirm if this file should be committed to the repo or if it should be in _.gitignore_?

thescientist13 commented on pom.xml:
```diff
@@ -0,0 +1,54 @@
+	<modelVersion>4.0.0</modelVersion>
+
+	<groupId>com.example</groupId>
+	<artifactId>test</artifactId>
```
@mdiwale 
we should change this to fit our project
- groupId: com.providencegeeks
- artifactId: website-api-posts

thescientist13 commented:
@mdiwale 
there are files in the _target/_ directory that are committed.  I believe _target/_ is the default output directory for Maven builds / artifacts, right?  If so, any build generated files, like these should not be tracked in version control.

You will need to:
1. remove them from being tracked by git
2. add _target/_ to .gitignore

mdiwale commented:
1) Updated README file
2) Added setup steps and other information about files in WIKI
3) Ignored Target
4) Changed the groupID and artifactID
5) .iml file -> IML is a module file created by IntelliJ IDEA, an IDE used to develop Java applications. It stores information about a development module, which may be a Java, Plugin, Android, or Maven component; saves the module paths, dependencies, and other settings.

Tried ignoring this file and still the program runs successfully. Hence it can be ignored.
Added test.iml file in ignore folder.

mdiwale commented:
@thescientist13 
Regarding action items, updated Wiki section about Java (version 8) and Maven installations.
Kindly let me know if any changes I need to make.

thescientist13 commented:
thanks @mdiwale.  

It looks good (just made some minor formatting edits). I am curious on the inclusion of Maven installation steps.  As we are providing Maven wrappers, there really isn't much need for contributors to install Maven, is there?

Also, in general, when providing installation instructions, if the tool provides good documentation for installing itself for Windows / OSX / Linux, etc, we should try and just leverage those docs and just link out to them instead.

Also, for the `mvn` commands, thanks for including those in the Onboarding guide, but I was thinking there might be project specific commands that used `spring-boot`.  Could you elaborate on what the differences are and when the `spring-boot` commands would be used vs without?

mdiwale commented:
@thescientist13 

Yes we are providing Maven Wrappers, hence no need to install Maven. But when I tried to run the application through command line using the command `mvn spring-boot:run`, it gave me an error for which I had to install Maven using `brew install maven` , after which my application executed successfully. So I just gave the steps if in case anyone needs it. I can remove it if you feel there is no need for it. 

Also, in general, when providing installation instructions, if the tool provides good documentation for installing itself for Windows / OSX / Linux, etc, we should try and just leverage those docs and just link out to them instead.  -> Should I remove everything and just provide the link??

thescientist13 commented:
Thanks @mdiwale !  

Here's my feedback
> Yes we are providing Maven Wrappers, hence no need to install Maven. But when I tried to run the application through command line using the command mvn spring-boot:run, it gave me an error for which I had to install Maven using brew install maven , after which my application executed successfully. So I just gave the steps if in case anyone needs it. I can remove it if you feel there is no need for it.

Then let's figure out what is needed then.  My understanding with providing Maven Wrapper is that contributor's _don't_ need to manage installing `mvn` locally.  If they still need to install it, what is the value of having the wrapper?  Let's figure out exactly what should be supported as a development workflow
1. `mvn`

or

2. Maven Wrapper

> Also, in general, when providing installation instructions, if the tool provides good documentation for installing itself for Windows / OSX / Linux, etc, we should try and just leverage those docs and just link out to them instead. -> Should I remove everything and just provide the link??

Yes, for instance when instructing users to install NodeJS, you can see how it was done for the [Frontend section](https://github.com/ProvidenceGeeks/website-docs/wiki/Onboarding-Guide#frontend) of the wiki docs, which just has a link to the NodeJS download page.


Hope that helps!!

mdiwale commented:
@thescientist13 

My understanding regarding Maven Wrapper:
When having a Maven Wrapper for a project there is no need to install a certain version of Maven on a machine.

We need to use `./mvnw clean package` and `./mvnw Spring-boot:run` to run the application and rest all the mvn commands (mentioned in [here](https://github.com/ProvidenceGeeks/website-docs/wiki/Onboarding-Guide#2-how-to-install-maven) ) can be used substituting mvn with ./mvnw

thescientist13 commented:
thanks for clarifying @mdiwale, that makes sense to me.  

In the [README](https://github.com/ProvidenceGeeks/website-api-events/pull/7/files#diff-04c6e90faac2675aa89e2176d2eec7d8R11) then it seems the test command should be changed to use the maven wrapper, correct?
e.g.
> To test everything works run `./mvnc spring-boot:run`

mdiwale commented:
Yes that's correct.

mdiwale commented:
@thescientist13 
Made the required changes !
